Add MySQL topology server implementation

morgo opened the pull request:
## Description

This adds a mysql topo server, where clients act like they are mysql replicas to subscribe to updates.

I used the vitess packages for binlog subscription/MySQL connectivity, which reduces code dependencies.

The local examples have been modified to test this, but due to #18355 it isn't being tested (yet).

I can add docs after it merges. It's pretty straightforward though.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/18340

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

morgo commented:
> @morgo would you be so kind as to create separate issues with specifics about those missing functionalities in our MySQL driver/connection implementation?

I managed to get the built in Vitess binlog library to work, provided that I require GTIDs. I'm fine with that limitation.

> There's another thing if you could explain please, about the notification system. I notice in file.go that any change (Create, Update, ...) invokes a ns.notifyChange. But this is only relevant to local process changes, right? What if a write is made from one tablet, and a 2nd tablet (or a vtgate, ...) is the watcher?

I've modified this. I'd intended to have it notify from both, but I've removed the change notification directly from the updates to ensure it only comes through replication. It looks to be working now.

shlomi-noach commented:
Thank you @morgo , a small request, please do not `force push` as this make it difficult to impossible to  track the changes done since my last review, especially given how large this change is 🙏

shlomi-noach commented on go/vt/topo/mysqltopo/lock.go:
```diff
@@ -0,0 +1,244 @@
+
+	// Check if the directory exists by looking for any files under it
+	var exists bool
+	likePattern := fullPath + "%"
```
I'm not sure whether this was intentionally resolved or as part of the force-push. It's a bit difficult to track the changes here, but I see the introduction of `func createLikePattern(prefix string) string {`. As far as I can see the question here still stands.`pattern += "%"` -- Is this going to be a problem if there is a lock called `foo` and we try to lock a new path called `foot`?

shlomi-noach commented on go/vt/topo/mysqltopo/lock.go:
```diff
@@ -0,0 +1,244 @@
+		}
+
+		// Check if it's a duplicate key error (lock already exists)
+		if isDuplicateKeyError(err) {
```
I see `INSERT IGNORE` was implemented.

shlomi-noach commented on go/vt/topo/mysqltopo/election.go:
```diff
@@ -0,0 +1,385 @@
+				return true // We renewed our leadership
+			}
+		}
+		log.Infof("Failed to renew leadership for %s (id: %s): %v", lp.name, lp.id, err)
```
(unresolving)

shlomi-noach commented on go/vt/topo/mysqltopo/lock.go:
```diff
@@ -0,0 +1,244 @@
+		return false
+	}
+	errStr := err.Error()
+	return strings.Contains(errStr, "Duplicate entry") || strings.Contains(errStr, "duplicate key")
```
I can't find the context after the forced push. How was this resolved? (unresolving)

shlomi-noach commented on go/vt/topo/mysqltopo/server_test.go:
```diff
@@ -0,0 +1,248 @@
+	// Wait for leadership with a longer timeout
+	leaderCtx, err := lp.WaitForLeadership()
+	if err != nil {
+		t.Logf("Failed to get leadership: %v", err)
```
The file now has 168 lines, where my comment was on line 226. I've lost context due t othe forced push. How was this resolved? Where has the code moved to? (I cannot find it). Unresolving.

morgo replied:
I removed this test, since it was covered by the topo tests (see `server_test.go:TestMySQLTopo()` which covers most of the testing). I have since added some additional election tests in election_test.go which covers some mysql specific cases I don't think are handled well in the topo tests.

This is things like concurrent requests in elections and renewing leadership, since that is built in in the other topos but has to be implemented in the MySQL case.

shlomi-noach commented on go/vt/topo/mysqltopo/watch.go:
```diff
@@ -0,0 +1,151 @@
+
+	// Create the watcher
+	watchCtx, cancel := context.WithCancel(ctx)
+	changes_chan := make(chan *topo.WatchData, 10) // Buffered channel
```
I see this was changed.

shlomi-noach commented on go/vt/topo/mysqltopo/notification.go:
```diff
@@ -0,0 +1,580 @@
+
+	// Notify watchers with the latest data
+	ns.notifyChange(path, data, MySQLVersion(version))
+}
```
I've lost the context to this after forced push. I cannot find neither the original code nor the auggested alternative. Can you please point me to the alternate code? Unresolving.

morgo replied:
I removed this code in the update:

- Previous behavior: an update/delete would also call notifyXX() so that the caller would have notification immediately.
- New behavior: it relies exclusively on detecting the change via the binlog.

shlomi-noach commented on go/vt/topo/mysqltopo/notification.go:
```diff
@@ -0,0 +1,626 @@
+				if ns.ctx.Err() != nil {
+					return // Context cancelled
+				}
+				log.Warningf("Binlog stream error: %v", err)
+				time.Sleep(time.Second)
+				continue
```
This may be an unrecoverable error, but it is silently dropped. The assumption is that we can `continue`, bu tthat is not guaranteed. We should error out, exit the function, temrinate the binlog connection, and retry/renew binlog connection fro the same position of interruption.

shlomi-noach replied:
Will returning an error retry the binlog connection?

morgo replied:
This looks to be a behavior difference between replication libraries. You are right - it doesn't try to handle and retry any errors in `StartBinlogDumpFromXX`.

I will change it so it has more robust handling.

shlomi-noach commented on go/vt/topo/mysqltopo/notification.go:
```diff
@@ -0,0 +1,626 @@
+		return
+	}
+
+	// For other events (INSERT, UPDATE, DELETE, QUERY etc)
+	// We treat them all the same which is to say that we
+	// check for changes. We don't try and understand the row image,
+	// because it requires tablemap parsing, and more complex logic.
+	ns.checkForTopoDataChanges()
```
This implies we're going to re-scan the `topo_data` table for any DML, even if unrelated, as well as for a myriad of other event types such as heartbeats, DDL, etc.

Is this desired and acceptable? It's an open question. IF the dataset is very small, and all this MySQL server ever does is to read/write `topo_data` table then this makes sense (but I'd still handle a bunch of other potential events). If this MySQL server happens to be doing even a little bit more, then why not validate that the event actually oeperates on `topo_data` table beforehand?

shlomi-noach replied:
The way to do that would be:

```go
case ev.IsTableMap():
   tm, err := ev.TableMap(format)
...
See:

https://github.com/vitessio/vitess/blob/8d90d26df84867f37570afd1bf6ebd80850bc9f8/go/mysql/binlog_event.go#L214-L215

Then, the next:
```go
case ev.IsWriteRows(), ev.IsUpdateRows(), ev.IsDeleteRows():
```
indicates a write on the table. But really that's not even needed, since the mere existence of TableMap with `tm.Name` being `"topo_data"` is enough.
I believe it's worth adding.

morgo commented:
I've pushed a few changes to the branch:
- The `WaitForLeadership()` implementation was working, but it could take up to timeout (30s) to detect leadership. There's no timeout in the etcd implementation, so I've rewritten it to something more compatible.
- I have added some MySQL specific election tests.
- I renamed `s.fullPath()` to `s.resolvePath()` and `createLikePattern()` to `matchDirectory()` - the handing is similar to what I had originally. It needs to match open ended wildcards. I have consistent tests to document it.
- The MySQL replication system retries connections if it disconnects. I have some basic resiliency testing for this, where it kills the MySQL binlog connection.
- The system that subscribes to MySQL events verifies that it matches the topo_data table.
- I check that the binary log format is RBR.
- I added back TestMySQLTopo() which I accidentally removed - this tests all the built in topo tests.

morgo commented on go.mod:
```diff
@@ -175,7 +175,7 @@ require (
 	github.com/go-logr/stdr v1.2.2 // indirect
 	github.com/go-ole/go-ole v1.2.6 // indirect
-	github.com/go-viper/mapstructure/v2 v2.3.0 // indirect
+	github.com/go-viper/mapstructure/v2 v2.2.1 // indirect
```
Thanks, I can revert.

shlomi-noach reviewed (commented):
Looking good and a couple minor styling nits.

I have not commented on my earlier review, but this PR warrants an endtoend test. What I'd expect is:
- Setting up a cluster with a MySQL topo backend
- Looking to see that the cluster starts up and is healthy
- Try reading from topo
- Try writing to topo
- Watch a topo entry (e.g. tablets)
- Make a change (e.g. change table type)
- Expect watch to work
- Anything else?
